File: leetcode/week_1_sprint/two_sum_hash_map_fundamentals/easy/two_sum_iv_input_is_a_bst/Solution.py
from collections import deque
from typing import Set, Optional, Deque, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int=0, left: Optional["TreeNode"]=None, right: Optional["TreeNode"]=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        # BFS and HashSet
        # Time Complexity O(n)
        # Space Complexity O(n)
        ## return self.approach_1(root, k)


        # DFS and HashSet
        # Time Complexity O(n)
        # Space Complexity O(n)
        ## return self.approach_2(root, k)

        # BST in-order traversal
        # Time Complexity O(n)
        # Space Complexity O(n)
        return self.approach_3(root, k)
    
    def approach_1(self, root: Optional[TreeNode], k: int) -> bool:
        seen: Set[int] = set()

        def dfs(node: Optional[TreeNode]) -> bool:
            if (not node):
                return False

            complement: int = k - node.val

            if (complement in seen):
                return True
            
            seen.add(node.val)

            return dfs(node.left) | dfs(node.right)
        
        return dfs(root)
    
    def approach_2(self, root: Optional[TreeNode], k: int) -> bool:
        seen: Set[int] = set()
        queue: Deque[Optional[TreeNode]] = deque([root])

        while (queue):
            node: Optional[TreeNode] = queue.popleft()
            if (not node):
                continue
            
            complement: int = k - node.val

            if (complement in seen):
                return True
            
            seen.add(node.val)

            queue.extend([node.left, node.right])
        
        return False

    def approach_3(self, root: Optional[TreeNode], k: int) -> bool:
        nodes_sorted: List[int] = []

        def inorder(node: Optional[TreeNode]) -> None:
            if (not node):
                return
            
            inorder(node.left)
            nodes_sorted.append(node.val)
            inorder(node.right)

            return
        
        inorder(root)
        
        low: int = 0
        high: int = len(nodes_sorted) - 1

        while (low < high):
            sum: int = nodes_sorted[low] + nodes_sorted[high]
            if (sum == k):
                return True
            elif (sum > k):
                high -= 1
            
            else:
                low += 1
            
        return False

File: leetcode/week_1_sprint/two_sum_hash_map_fundamentals/easy/two_sum_iv_input_is_a_bst/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        // HashSet + DFS
        // Time Complexity O(n)
        // Space Complexity O(n)
        // return approach_1(root, k);

        // HashSet + BFS
        // Time Complexity O(n)
        // Space Complexity O(n)
        // return approach_2(root, k);

        // BST in-order traversal
        // Time Complexity O(n)
        // Space Complexity O(n)
        return approach_3(root, k);
    }

    bool approach_1(TreeNode* root, int k) {
        std::unordered_set<int> uset;

        std::function<bool(TreeNode*)> dfs = [&](TreeNode* node) {
            if (node == nullptr) {
                return false;
            }
            int complement = k - node->val;

            if (uset.count(complement) == 1) {
                return true;
            }
            uset.insert(node->val);

            return dfs(node->left) || dfs(node->right);
        };

        return dfs(root);
    }

    bool approach_2(TreeNode* root, int k) {
        std::unordered_set<int> uset;
        std::deque<TreeNode*> dque;

        dque.push_back(root);

        while (!dque.empty()) {
            TreeNode* node = dque.front();
            dque.pop_front();
            if (node == nullptr) {
                continue;
            }

            int complement = k - node->val;
            if (uset.count(complement) == 1) {
                return true;
            }

            uset.insert(node->val);

            dque.push_back(node->left);
            dque.push_back(node->right);
        }

        return false;
    }

    bool approach_3(TreeNode* root, int k) {
        std::vector<int> v;

        std::function<void(TreeNode*)> in_order = [&](TreeNode* node) {
            if (node == nullptr) return;

            in_order(node->left);
            v.push_back(node->val);
            in_order(node->right);
        };

        in_order(root);

        int low = 0;
        int high = v.size() - 1;

        while (low < high) {
            int sum = v[low] + v[high];
            if (sum == k) {
                return true;
            }

            if (sum > k) {
                --high;
            }

            else {
                ++low;
            }
        }

        return false;

    }
}